core-primitives/stf-state-handler/src/file_io.rs

```diff
@@ -18,134 +18,379 @@
+	) -> Result<Self::HashType>;
+
+	/// Create an empty (default initialized) state.
+	fn create_initialized(
+		&self,
+		shard_identifier: &ShardIdentifier,
+		state_id: StateId,
+	) -> Result<Self::HashType>;
+
+	/// Write the state.
+	fn write(
+		&self,
+		shard_identifier: &ShardIdentifier,
+		state_id: StateId,
+		state: Self::StateType,
+	) -> Result<Self::HashType>;
+
+	/// Remove a state.
+	fn remove(&self, shard_identifier: &ShardIdentifier, state_id: StateId) -> Result<()>;
+
+	/// Checks if a given shard directory exists and contains at least one state instance.
+	fn shard_exists(&self, shard_identifier: &ShardIdentifier) -> bool;
+
+	/// Lists all shards.
+	fn list_shards(&self) -> Result<Vec<ShardIdentifier>>;
+
+	/// List all states for a shard.
+	fn list_state_ids_for_shard(&self, shard_identifier: &ShardIdentifier) -> Result<Vec<StateId>>;
 }
 
-/// Writes the state (without the state diff) encrypted into the enclave storage
-/// Returns the hash of the saved state (independent of the diff!)
-pub(crate) fn write(state: StfState, shard: &ShardIdentifier) -> Result<H256> {
-	let state_path =
-		format!("{}/{}/{}", SHARDS_PATH, shard.encode().to_base58(), ENCRYPTED_STATE_FILE);
-	trace!("writing state to: {}", state_path);
+#[cfg(feature = "sgx")]
+pub mod sgx {
+
+	use super::*;
+	use crate::state_key_repository::AccessStateKey;
+	use ita_stf::{State as StfState, StateType as StfStateType, Stf};
+	use itp_sgx_crypto::StateCrypto;
+	use itp_sgx_io::{read as io_read, write as io_write};
+	use itp_types::H256;
+	use log::*;
+	use sgx_tcrypto::rsgx_sha256_slice;
+	use std::{path::Path, sync::Arc};
+
+	/// SGX state file I/O.
+	pub struct SgxStateFileIo<StateKeyRepository> {
+		state_key_repository: Arc<StateKeyRepository>,
+	}
+
+	impl<StateKeyRepository> SgxStateFileIo<StateKeyRepository>
+	where
+		StateKeyRepository: AccessStateKey,
+	{
+		pub fn new(state_key_repository: Arc<StateKeyRepository>) -> Self {
+			SgxStateFileIo { state_key_repository }
+		}
+
+		fn read(&self, path: &Path) -> Result<Vec<u8>> {
+			let mut bytes = io_read(path)?;
+
+			if bytes.is_empty() {
+				return Ok(bytes)
+			}
+
+			let state_hash = rsgx_sha256_slice(&bytes)?;
+			debug!(
+				"read encrypted state with hash {:?} from {:?}",
+				H256::from_slice(state_hash.as_ref()),
+				path
+			);
+
+			let state_key = self.state_key_repository.retrieve_key()?;
+
+			state_key
+				.decrypt(&mut bytes)
+				.map_err(|e| Error::Other(format!("{:?}", e).into()))?;
+			trace!("buffer decrypted = {:?}", bytes);
+
+			Ok(bytes)
+		}
+
+		fn encrypt(&self, mut state: Vec<u8>) -> Result<Vec<u8>> {
+			let state_key = self.state_key_repository.retrieve_key()?;
+
+			state_key
+				.encrypt(&mut state)
+				.map_err(|e| Error::Other(format!("{:?}", e).into()))?;
+			Ok(state)
+		}
+	}
+
+	impl<StateKey> StateFileIo for SgxStateFileIo<StateKey>
+	where
+		StateKey: AccessStateKey,
+	{
+		type StateType = StfState;
+		type HashType = H256;
+
+		fn load(
```

Comment:
```suggestion
		fn load_and_initialize(
```

Reply:
or mabye `load_or_initialize`

Reply:
It does not initialize anything, it only loads. Returns an error if you try to load something that does not exist.

Reply:
What does this part do then? https://github.com/integritee-network/worker/blob/c06dcd6b98f6138b582252972a3f5d11d3692e08/core-primitives/stf-state-handler/src/file_io.rs#L171-L174

Reply:
You are right! I think this part is a relic from the times when this function did initialize. I'll remove it 👍